test/functional/p2p_invalid_tx.py

```diff
@@ -44,12 +65,48 @@ def run_test(self):
 
         # b'\x64' is OP_NOTIF
         # Transaction will be rejected with code 16 (REJECT_INVALID)
+        self.log.info('Test a transaction that is rejected')
         tx1 = create_transaction(block1.vtx[0], 0, b'\x64', 50 * COIN - 12000)
-        node.p2p.send_txs_and_test([tx1], node, success=False, reject_code=16, reject_reason=b'mandatory-script-verify-flag-failed (Invalid OP_IF construction)')
+        node.p2p.send_txs_and_test([tx1], node, success=False, expect_disconnect=True)
+        self.reconnect_p2p(additional_conn=True)  # Add one more (that is later disconnected)
+
+        self.log.info('Test orphan transaction handling ... ')
+        tx_withhold = CTransaction()
+        tx_withhold.vin.append(CTxIn(outpoint=COutPoint(block1.vtx[0].sha256, 0)))
+        tx_withhold.vout.append(CTxOut(nValue=50 * COIN - 12000, scriptPubKey=b'\x51'))
+        tx_withhold.calc_sha256()
+
+        tx_orphan_1 = CTransaction()
+        tx_orphan_1.vin.append(CTxIn(outpoint=COutPoint(tx_withhold.sha256, 0)))
+        tx_orphan_1.vout = [CTxOut(nValue=10 * COIN, scriptPubKey=b'\x51')] * 3
+        tx_orphan_1.calc_sha256()
+
+        tx_orphan_2_no_fee = CTransaction()
+        tx_orphan_2_no_fee.vin.append(CTxIn(outpoint=COutPoint(tx_orphan_1.sha256, 0)))
+        tx_orphan_2_no_fee.vout.append(CTxOut(nValue=10 * COIN, scriptPubKey=b'\x51'))
+
+        tx_orphan_2_valid = CTransaction()
+        tx_orphan_2_valid.vin.append(CTxIn(outpoint=COutPoint(tx_orphan_1.sha256, 1)))
+        tx_orphan_2_valid.vout.append(CTxOut(nValue=10 * COIN - 12000, scriptPubKey=b'\x51'))
+        tx_orphan_2_valid.calc_sha256()
+
+        tx_orphan_2_invalid = CTransaction()
+        tx_orphan_2_invalid.vin.append(CTxIn(outpoint=COutPoint(tx_orphan_1.sha256, 2)))
+        tx_orphan_2_invalid.vout.append(CTxOut(nValue=11 * COIN, scriptPubKey=b'\x51'))
+
+        self.log.info('Send the orphans ... ')
+        # Send orphan txs that are valid by consensus, but may be rejected by policy from p2ps[0]
+        node.p2p.send_txs_and_test([tx_orphan_1, tx_orphan_2_no_fee, tx_orphan_2_valid], node, success=False)
+        # Send consensus-invalid tx from p2ps[1], they should get disconnected later on
+        node.p2ps[1].send_txs_and_test([tx_orphan_2_invalid], node, success=False)
+
+        self.log.info('Send the withhold tx ... ')
+        assert_equal(2, len(node.getpeerinfo()))
+        # The tx itself is successfully accepted, but we should get disconnected anyway
+        node.p2p.send_txs_and_test([tx_withhold], node, success=True, expect_disconnect=False)
+        assert_equal(1, len(node.getpeerinfo()))
 
-        # Verify valid transaction
-        tx1 = create_transaction(block1.vtx[0], 0, b'', 50 * COIN - 12000)
-        node.p2p.send_txs_and_test([tx1], node, success=True)
+        assert_equal({t.hash for t in [tx_withhold, tx_orphan_1, tx_orphan_2_valid]}, set(node.getrawmempool()))
```

Comment:
Could make sense to rebase the test commit before the refactoring commit, to make it clear existing code is supposed to pass the new test.

Or maybe just put the new test in a standalone PR if the refactoring is going to be held up by #11775 anyway (https://github.com/bitcoin/bitcoin/pull/12935#discussion_r180830562).